docs/core/tutorials/testing-library-with-visual-studio-code.md

```diff
@@ -0,0 +1,187 @@
+| `Assert.IsNotNull` | Verifies that an object isn't `null`. The assert fails if the object is `null`. |
+
+You can also use the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.Assert.ThrowsException%2A?displayProperty=nameWithType> method in a test method to indicate the type of exception it's expected to throw. The test fails if the specified exception isn't thrown.
+
+In testing the `StringLibrary.StartsWithUpper` method, you want to provide a number of strings that begin with an uppercase character. You expect the method to return `true` in these cases, so you can call the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.Assert.IsTrue%2A?displayProperty=nameWithType> method. Similarly, you want to provide a number of strings that begin with something other than an uppercase character. You expect the method to return `false` in these cases, so you can call the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.Assert.IsFalse%2A?displayProperty=nameWithType> method.
+
+Since your library method handles strings, you also want to make sure that it successfully handles an [empty string (`String.Empty`)](xref:System.String.Empty), a valid string that has no characters and whose <xref:System.String.Length> is 0, and a `null` string that hasn't been initialized. If `StartsWithUpper` is called as an extension method on a <xref:System.String> instance, it can't be passed a `null` string. However, you can also call it directly as a static method and pass a single <xref:System.String> argument.
+
+You'll define three methods, each of which calls an <xref:Microsoft.VisualStudio.TestTools.UnitTesting.Assert> method repeatedly for each element in a string array. Because the test method fails as soon as it finds the first failure, you'll call a method overload that allows you to pass a string that indicates the string value used in the method call.
+
+To create the test methods:
+
+1. Open *StringLibraryTest/UnitTest1.cs* and replace all of the code with the following code.
+
+   :::code language="csharp" source="./snippets/library-with-visual-studio/csharp/StringLibraryTest/UnitTest1.cs":::
+
+   The test of uppercase characters in the `TestStartsWithUpper` method includes the Greek capital letter alpha (U+0391) and the Cyrillic capital letter EM (U+041C). The test of lowercase characters in the `TestDoesNotStartWithUpper` method includes the Greek small letter alpha (U+03B1) and the Cyrillic small letter Ghe (U+0433).
+
+1. Save your changes.
+
+1. Run the tests:
+
+   ```dotnetcli
+   dotnet test StringLibraryTest/StringLibraryTest.csproj
+   ```
+
+   The terminal output shows that all tests passed.
+
+   ```
+   Starting test execution, please wait...
+
+   A total of 1 test files matched the specified pattern.
+
+   Test Run Successful.
+   Total tests: 3
+        Passed: 3
+   Total time: 5.1116 Seconds
+   ```
+
+## Handle test failures
+
+Your test run had no failures, but change it slightly so that one of the test methods fails:
```

Comment:
It might be wise to add text here to explain the importance of this technique when writing tests for existing code (that was written without tests).